### PR TITLE
Add the possibility to search a pokemon

### PR DESCRIPTION
**Pull request by Matttx**
Hi again !

I add the possibility to search a Pokemon with the Pokedex SearchBar

Simulator Screen Shot - iPhone 13 - 2021-11-24 at 20 24 11

### CODE REVIEW

**Comment by Hugovidafe** [comment]
I searched pokemons without download all images from the pokemon list.
And there ocurred an exception.

image

**Comment by Matttx** [comment]
I trying to see how to avoid the `innerController?.dispose();` when we are searching a Pokemon but I don't find anything 😕

If someone know how to do that, pls explain me haha

File: lib/data/repositories/pokemon_repository.dart [hunk 1]
```diff
@@ -16,6 +18,27 @@ class PokemonDefaultRepository extends PokemonRepository {
...
+  @override
+  Future<List<Pokemon>> searchPokemon(String pokemonName) async {
+    final hasCachedData = await localDataSource.hasData();
+
+    if (!hasCachedData) {
+      final pokemonGithubModels = await githubDataSource.getPokemons();
+      final pokemonHiveModels = pokemonGithubModels.map((e) => e.toHiveModel());
+
+      await localDataSource.savePokemons(pokemonHiveModels);
+    }
```
Comment (scitbiz) [inline lib/data/repositories/pokemon_repository.dart:30]: This one use in multiple places, you should break it into a new method, something like `fetchAndCachePokemons`

File: lib/data/source/local/local_datasource.dart [hunk 2]
```diff
@@ -35,6 +35,16 @@ class LocalDataSource {
     await pokemonBox.putAll(pokemonsMap);
   }
 
+  Future<List<PokemonHiveModel>> searchPokemon({String pokemonName}) async {
+    final pokemonBox = Hive.box<PokemonHiveModel>(PokemonHiveModel.boxKey);
+
+    final pokemons = List.generate(pokemonBox.length, (index) => pokemonBox.getAt(index));
+
+    final filteredPokemons = pokemons.where((pokemon) => pokemon.name.contains(pokemonName)).toList();
```
Comment (scitbiz) [inline lib/data/source/local/local_datasource.dart:43]: use `toLowerCase()` for both `pokemonName` and `pokemon.name` to allow user to search by both uppercase and lowercase

File: lib/states/pokemon_list_state.dart [hunk 3]
```diff
@@ -3,16 +3,33 @@ part of 'states.dart';
...
   List<Pokemon> pokemons = [];
 
+  void searchPokemon(String pokemonName) async {
+    isError = false;
+    loading = true;
+    notifyListeners();
+
+    try {
+      pokemons = await _searchPokemonUseCase(
+          SearchPokemonParams(pokemonName: pokemonName));
```
Comment (scitbiz) [inline lib/states/pokemon_list_state.dart:24]:
```suggestion
          SearchPokemonParams(pokemonName: pokemonName),);
```
Please remember to add the little `,` to make the formatter working nicely

File: lib/ui/modals/search_modal.dart [hunk 4]
```diff
@@ -18,7 +22,11 @@ class SearchBottomModal extends StatelessWidget {
             top: context.responsive(14),
             bottom: context.responsive(14) + viewInsets + safeAreaBottom,
           ),
-          child: SearchBar(margin: EdgeInsets.zero),
+          child: SearchBar(margin: EdgeInsets.zero, onSubmit: (content) {
+              scheduleMicrotask(() {
```
Comment (scitbiz) [inline lib/ui/modals/search_modal.dart:26]: What is `scheduleMicrotask` doing in this case?

**Review by scitbiz — CHANGES_REQUESTED** [review]
@Matttx Thanks for making this pull request!!

**Comment by scitbiz** [comment]
@Matttx Are you interested in adding the search result in the future too? That would be awesome!

File: lib/ui/modals/search_modal.dart (hunk 4 above)
Comment (Matttx) [reply lib/ui/modals/search_modal.dart:26]: Oups, that was a copy paste mistake sorry 😅

**Comment by Matttx** [comment]
> @Matttx Are you interested in adding the search result in the future too? That would be awesome!

It's depending on my schedule, because I have a lot of work to do in this moment, but the project looks really good and It's a pleasure to work on it. Thank you for the proposition 😄

**Comment by scitbiz** [comment]
@Matttx Ah it's okay, I'm currently in the same position as yours. It's better if you're the one who will finish the whole task.
I will leave the task for you if you're planning to do it yourself. But if you're busy, I will happy to do it or maybe any other contributor wants to do it too :D

**Comment by scitbiz** [comment]
@Matttx I think the result should be shown inside the search bottom sheet. That means we need to create a separate state for that and not use the `pokemon_list` state. And the above bug will not occurred anymore

File: lib/data/repositories/pokemon_repository.dart [hunk 5]
```diff
@@ -5,6 +5,8 @@ import 'package:pokedex/data/source/mappers/local_to_entity_mapper.dart';
 import 'package:pokedex/domain/entities/pokemon.dart';
 
 abstract class PokemonRepository {
+  Future<List<Pokemon>> searchPokemon(String pokemon);
```
Comment (Hugovidafe) [inline lib/data/repositories/pokemon_repository.dart:8]:
```suggestion
  Future<List<Pokemon>> searchPokemons(String pokemon);
```

File: lib/data/repositories/pokemon_repository.dart [hunk 6]
```diff
@@ -26,6 +28,26 @@ class PokemonDefaultRepository extends PokemonRepository {
 
       await localDataSource.savePokemons(pokemonHiveModels);
     }
+  }
+
+  @override
+  Future<List<Pokemon>> searchPokemon(String pokemonName) async {
```
Comment (Hugovidafe) [inline lib/data/repositories/pokemon_repository.dart:34]:
```suggestion
  Future<List<Pokemon>> searchPokemons(String pokemonName) async {
```

File: lib/data/repositories/pokemon_repository.dart [hunk 7]
```diff
@@ -26,6 +28,26 @@ class PokemonDefaultRepository extends PokemonRepository {
...
+    await _fetchAndCachePokemons(githubDataSource, localDataSource);
+
+    final pokemonHiveModels =
+        await localDataSource.searchPokemon(pokemonName: pokemonName);
```
Comment (Hugovidafe) [inline lib/data/repositories/pokemon_repository.dart:38]:
```suggestion
        await localDataSource.searchPokemons(pokemonName: pokemonName);
```

File: lib/data/source/local/local_datasource.dart [hunk 8]
```diff
@@ -35,6 +35,16 @@ class LocalDataSource {
     await pokemonBox.putAll(pokemonsMap);
   }
 
+  Future<List<PokemonHiveModel>> searchPokemon({String pokemonName}) async {
```
Comment (Hugovidafe) [inline lib/data/source/local/local_datasource.dart:38]:
```suggestion
  Future<List<PokemonHiveModel>> searchPokemons({String pokemonName}) async {
```

File: lib/domain/usecases/pokemon_usecases.dart [hunk 9]
```diff
@@ -4,6 +4,24 @@ import '../../core/usecase.dart';
...
+}
+
+class SearchPokemonUseCase extends UseCase<List<Pokemon>, SearchPokemonParams> {
+  const SearchPokemonUseCase(this.repository);
+
+  final PokemonRepository repository;
+
+  @override
+  Future<List<Pokemon>> call(SearchPokemonParams params) {
+    return repository.searchPokemon(params.pokemonName);
```
Comment (Hugovidafe) [inline lib/domain/usecases/pokemon_usecases.dart:20]:
```suggestion
    return repository.searchPokemons(params.pokemonName);
```

File: lib/ui/modals/search_modal.dart [hunk 10]
```diff
@@ -18,7 +20,9 @@ class SearchBottomModal extends StatelessWidget {
             top: context.responsive(14),
             bottom: context.responsive(14) + viewInsets + safeAreaBottom,
           ),
-          child: SearchBar(margin: EdgeInsets.zero),
+          child: SearchBar(margin: EdgeInsets.zero, onSubmit: (content) {
+            context.read(pokemonsStateProvider).searchPokemon(content);
```
Comment (Hugovidafe) [inline lib/ui/modals/search_modal.dart:24]:
```suggestion
            context.read(pokemonsStateProvider).searchPokemons(content);
```

**Review by Hugovidafe — COMMENTED** [review]
The search can be used to search for multiple Pokemons; for example, searching for "chu", there are several Pokemons...
It's just a suggestion ^^

**Comment by Hugovidafe** [comment]
> @Matttx I think the result should be shown inside the search bottom sheet. 

I would prefer to put the search engine at the bottom, so as not to have to duplicate another state of list_pokemon

Only a suggestion ^^

**Comment by scitbiz** [comment]
@Hugovidafe That is also nice too, but I want to keep the app look the same as the design so I think the best option is to keep the overall list and the search result separately. And separating things make the code easier to maintain too

**Comment by Matttx** [comment]
> The search can be used to search for multiple Pokemons; for example, searching for "chu", there are several Pokemons...
> It's just a suggestion ^^

It's already doing this, it's just a spelling mistake haha

**Comment by Matttx** [comment]
> @Hugovidafe That is also nice too, but I want to keep the app look the same as the design so I think the best option is to keep the overall list and the search result separately. And separating things make the code easier to maintain too

I'll do it 😁

**Comment by Matttx** [comment]
Do I create a new state for the search Pokemon feature or I just add a pokemons search list inside the PokemonListState ?

**Comment by scitbiz** [comment]
@Matttx I think you should create a new one because `PokemonListState` clearly defines what it is doing. And we also need to use another `loading` and `error` for it so that the main list doesn't show the loading icon, right?

**Comment by scitbiz** [comment]
Hi @Matttx, is this PR ready for review yet?
Please turn this PR to `Ready for review` whenever you are ready!

**Pull request closed without merging**